backend/ipfs/filemanager.go

```diff
@@ -0,0 +1,301 @@
+package ipfs
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/ipfs/boxo/blockservice"
+	blockstore "github.com/ipfs/boxo/blockstore"
+	chunker "github.com/ipfs/boxo/chunker"
+	"github.com/ipfs/boxo/exchange"
+	"github.com/ipfs/boxo/files"
+	"github.com/ipfs/boxo/ipld/merkledag"
+	unixfile "github.com/ipfs/boxo/ipld/unixfs/file"
+	"github.com/ipfs/boxo/ipld/unixfs/importer/balanced"
+	"github.com/ipfs/boxo/ipld/unixfs/importer/helpers"
+	"github.com/ipfs/boxo/provider"
+	"github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+	multihash "github.com/multiformats/go-multihash"
+	"go.uber.org/zap"
+)
+
+const (
+	// IPFSRootRoute is the root route of ipfs.
+	IPFSRootRoute = "/ipfs"
+	// UploadRoute is the route to upload a file.
+	UploadRoute = "/file-upload"
+	routeVar    = "cid"
+	// GetRoute is the route to get a file.
+	GetRoute = "/{" + routeVar + "}"
+	// MaxFileMB is the maximum file size (in MB) to be uploaded.
+	MaxFileMB = 64
+	// SearchTimeout is the maximum time we are searching for a file.
+	SearchTimeout = 2 * time.Minute
+)
+
+// AddParams contains all of the configurable parameters needed to specify the
+// importing process of a file.
+type AddParams struct {
+	Layout    string
+	Chunker   string
+	RawLeaves bool
+	Hidden    bool
+	Shard     bool
+	NoCopy    bool
+	HashFun   string
+}
+
+// HTTPHandler is an interface to pass to the router only the http handlers and
+// not all the FileManager type.
+type HTTPHandler interface {
+	GetFile(http.ResponseWriter, *http.Request)
+	UploadFile(http.ResponseWriter, *http.Request)
+}
+
+// FileManager is the main object to handle ipfs files.
+type FileManager struct {
+	ctx        context.Context
+	started    bool
+	log        *zap.Logger
+	exch       exchange.Interface
+	bstore     blockstore.Blockstore
+	bservice   blockservice.BlockService
+	DAGService ipld.DAGService // become a DAG service
+	provider   provider.System
+}
+
+// NewManager creates a new fileManager instance.
+func NewManager(ctx context.Context, log *zap.Logger) *FileManager {
+	return &FileManager{
+		log:     log,
+		ctx:     ctx,
+		started: false,
+	}
+}
+
+// Start starts new manager.
+func (fm *FileManager) Start(blockstore blockstore.Blockstore, bitswap *Bitswap, provider provider.System) error {
```

Comment:
@juligasa What does this `Start` do? Are there any goroutines in the background for handling file uploads?

Reply:
This has to do with the lazy initialization. When we call start all the params needed (blockstore, bitswap, etc) are available. When calling `New` instead, they are not. And we could not pass a future like in the other cases because we have a cyclic dependency loop between ifps and mttnet packages

Reply:
oooh, I see. Hope we could improve that soon.

Reply:
@juligasa could you add a comment somewhere to explain this?

Reply:
Sure!